### ISSUE TITLE
Test_Agent fails asserting that '3.6' is of type float

### ISSUE THREAD

**Issue by rickmacgillis** [fuel/core#1815]
PHP returns string based versions of all numbers from get_browser() and thus, Test_Agent fails to assert that '3.6' is of type float.

**Steps to Reproduce it**
1. Run \Debug::dump(\Agent::version());to see that a string version of the number is generated. (Run browscap through PHP, as FuelPHP fails to parse the 8MB ini file without tripping the memory limit. Lifting the memory limit doesn't appear to fix the situation.)
2. Run php oil test and see the following error message.
   1) Fuel\Core\Test_Agent::test_version
   Failed asserting that '3.6' is of type "float".
   /fuel/core/tests/agent.php:71
3. To verify where the issue comes from, run:
   `\Debug::dump(get_browser($_SERVER['HTTP_USER_AGENT'], true));`
   You'll notice that PHP's native get_browser() returns the string based version of the numbers.

Also, please note that Browscap is an active project and it could be due to the current Browscap ini files containing the string based version all all numbers, whereas previous versions were properly type cased. PHP probably doesn't cast these numbers to the float type as that would be bad practice for a programming language to alter external data like that.

**Proposed solution**
Consider removing the type check, as you already have a check to verify that the output is exactly the same. (assertEquals) You might also consider type checking for a string, but that could throw many false positives, and would be less productive.

UPDATE: Also, false values are reported as empty by get_browser(), so other checks that rely on them also fail.

**Comment by WanWizard** [comment]
I can not reproduce this. 

Tests run against an included known version of the browscap data, just to avoid these issues. Again, are you testing an old version of Fuel?

**Comment by rickmacgillis** [comment]
I'm running 1.7.2.

**Comment by WanWizard** [comment]
Can you re-run the same test in 1.8/develop? Because it could be fixed since the 1.7.2. release.

**Comment by rickmacgillis** [comment]
Alright, I've upgraded to 1.8 today. The problem still exists on both accounts. I get "Failed asserting that '3.6' is of type "float"." Also, I get "Failed asserting that '' is false." Both are described in detail in the reports. I've also checked the contents of the agent.browscap.file config value from the test and it states the correct file, as you mentioned. So we are testing off of a "known good" file. Let me do a clean install and see what happens. I'll report back in a second.

Update: I've tested this on a clean install and I still get the errors. I managed to figure out the cause of this mess. For anyone who comes across this issue, somehow FuelPHP gets confused when Browscap is natively installed in PHP. (Ex. browscap = /etc/php.d/browscap.ini in your php.ini file) I've turned that off and everything seems to be working well now.

**Comment by WanWizard** [comment]
Are you running tests in an application? Because it might be that some of your changes influence the result. On a clean install I can not reproduce it.

**Comment by rickmacgillis** [comment]
See the above update.

**Comment by WanWizard** [comment]
Ok, missed that. The agent class automatically switches to php's native get_browser() if the browscap file is installed, the test doesn't override that.

That means the test doesn't have a predictable environment again, something that needs to be looked at, because it also means you get different results, appearently get_browser() returns strings instead of a float (in this case).

**Comment by rickmacgillis** [comment]
I'd also like to point out that for anyone who's using \Agent for anything, Using the native get_browser() stuff makes FuelPHP run immensely faster. Only use the FuelPHP alternative is you really don't have any other way around it.

From looking at the code, I don't think there really is an easy way to completely speed up the FuelPHP version for processing a huge amount of data each time it runs. FuelPHP could cache the browser data values in a session, but it will need to use heavy resources each time it runs the file for each user, as long as they use the same browser, and don't clear their cookies. If the user updates their browser, then the system would be out of sync.

In short, FuelPHP is doing the best it can to process a 6MB+ file in a highly dynamic environment, but native PHP runs faster, as it's native. (It's meant to do what it says it does in a native environment.) I've added back in my native support.

Also, if you have special requirements, like in a shared hosting environment, see if your host allows you to use custom ini files, and set it up there.

**Comment by WanWizard** [comment]
As said before, Agent will automatically switch to get_browser() if it is defined in your php.ini. 

The fallback is only there for environments in which you don't have access to the ini file, for example in shared hosting. Since you normally don't use a shared hosting environment for anything serious, I don't see that as a problem. Agent's performance will be addressed in v2.
